iron-input.html

```diff
@@ -125,11 +169,52 @@
      * @suppress {checkTypes}
      */
     _bindValueChanged: function() {
+      if (typeof this.bindValue !== 'string') {
+        // it will trigger another _bindValueChanged event
+        this.bindValue = !(this.bindValue || this.bindValue === 0) ? '' : this.bindValue + '';
+      } else {
+        this._updateValueIfNeeded();
+        // manually notify because we don't want to notify until after setting value
+        this.fire('bind-value-changed', {value: this.bindValue});
+      }
+    },
+
+    _bindValueAsNumberChanged: function() {
+      // ensure it is a number
+      if (typeof this.bindValueAsNumber !== 'number') {
+        // it will trigger another _bindValueAsNumberChanged event
+        if (isNaN(this.bindValueAsNumber)) {
+          this.bindValueAsNumber = NaN;
+        } else {
+          this.bindValueAsNumber = parseFloat(this.bindValueAsNumber);
+        }
+      } else {
+        this._updateValueAsNumberIfNeeded();
+        // manually notify because we don't want to notify until after setting valueAsNumber
+        this.fire('bind-value-as-number-changed', {value: this.bindValueAsNumber});
```

Comment:
If you configure the `bindValueAsNumber` property with `notify: true`, I think it will do this for you as you update the value.